pkg/store/proxy.go

```diff
@@ -344,11 +398,55 @@ func startStreamSeriesSet(
 
 	wg.Add(1)
 	go func() {
+		isFirstRecv := true
 		defer wg.Done()
 		defer close(s.recvCh)
-
 		for {
-			r, err := s.stream.Recv()
+			var cancel context.CancelFunc
+			if s.responseTimeout != 0 {
+				ctx, cancel = context.WithTimeout(ctx, s.responseTimeout)
+				defer cancel()
+			}
+			rCh := make(chan *recvResponse, 1)
+			var rr *recvResponse
+			go func() {
+				t0 := time.Now()
+				r, err := s.stream.Recv()
+				if isFirstRecv {
+					if r.Size() > 0 {
+						metrics.withPayload.Observe(time.Since(t0).Seconds())
+					} else {
+						metrics.withoutPayload.Observe(time.Since(t0).Seconds())
+					}
+					isFirstRecv = false
+				}
+				rCh <- &recvResponse{r: r, err: err}
+			}()
+
+			select {
+			case <-ctx.Done():
```

Comment:
I think we might want to separate response timeout context vs the given context. Just want to avoid this when client cancels request or we have overall query timeout.